controlplane/kubeadm/internal/machinefilters/machine_filters.go

```diff
@@ -197,3 +205,114 @@ func ControlPlaneSelectorForCluster(clusterName string) labels.Selector {
 		must(labels.NewRequirement(clusterv1.MachineControlPlaneLabelName, selection.Exists, []string{})),
 	)
 }
+
+// MatchesKCPConfiguration returns a filter to find all machines that matches with KCP config and do not require any rollout.
+// Kubernetes version, infrastructure template, and KubeadmConfig field need to be equivalent.
+func MatchesKCPConfiguration(ctx context.Context, c client.Client, kcp controlplanev1.KubeadmControlPlane, cluster clusterv1.Cluster) func(machine *clusterv1.Machine) bool {
+	return And(
+		MatchesKubernetesVersion(kcp.Spec.Version),
+		MatchesKubeadmBootstrapConfig(ctx, c, kcp, cluster),
+		MatchesTemplateClonedFrom(ctx, c, kcp),
+	)
+}
+
+// MatchesTemplateClonedFrom returns a filter to find all machines that match a given KCP infra template.
+func MatchesTemplateClonedFrom(ctx context.Context, c client.Client, kcp controlplanev1.KubeadmControlPlane) Func {
+	return func(machine *clusterv1.Machine) bool {
+		if machine == nil {
+			return false
+		}
+		infraObj, err := external.Get(ctx, c, &machine.Spec.InfrastructureRef, kcp.Namespace)
+		// Return true here because failing to get infrastructure machine should not be considered as unmatching.
+		if err != nil {
+			return true
+		}
+
+		clonedFromName, ok1 := infraObj.GetAnnotations()[clusterv1.TemplateClonedFromNameAnnotation]
+		clonedFromGroupKind, ok2 := infraObj.GetAnnotations()[clusterv1.TemplateClonedFromGroupKindAnnotation]
+		if !ok1 || !ok2 {
+			// All kcp cloned infra machines should have this annotation.
+			// Missing the annotation may be due to older version machines or adopted machines.
+			// Should not be considered as mismatch.
+			return true
+		}
+
+		//Check if infra machine's InfrastructureTemplate is matching the kcp InfrastructureTemplate.
+		if clonedFromName != kcp.Spec.InfrastructureTemplate.Name ||
```

Comment:
Is it enough to check the infrastructure provider template name to assume it is not changed?
from the [book](https://cluster-api.sigs.k8s.io/tasks/change-machine-template.html) it seems infrastructure provider might make templates mutable...
```
Some infrastructure providers may, at their discretion, choose to support in-place modifications of certain infrastructure machine template fields.
```

Reply:
Last sentence in that page:

` In such cases, however, Cluster API will not trigger a rolling update.`

Is there any cases we will need to support?

Reply:
This should be enough, templates are treated as immutable references, they can be mutable if and only if the mutations are handled by infrastructure providers; Cluster API won't roll out new Machines for KCP or MachineSet unless you change the template reference.

Might be worthwhile to add some better description in the docs.